spring-integration-core/src/test/java/org/springframework/integration/endpoint/PollerAdviceTests.java

```diff
@@ -113,6 +135,109 @@ public boolean skipPoll() {
+		final List<String> callOrder = new ArrayList<String>();
+		final CountDownLatch latch = new CountDownLatch(4);// advice + advice + source + advice
+		adapter.setSource(new MessageSource<Object>() {
+
+			@Override
+			public Message<Object> receive() {
+				callOrder.add("c");
+				latch.countDown();
+				return null;
+			}
+		});
+		adapter.setTrigger(new Trigger() {
+
+			private boolean done;
+
+			@Override
+			public Date nextExecutionTime(TriggerContext triggerContext) {
+				Date date = done ? null : new Date(System.currentTimeMillis() + 10);
+				done = true;
+				return date;
+			}
+		});
+		configure(adapter);
+		List<Advice> adviceChain = new ArrayList<Advice>();
+
+		class TestGeneralAdvice implements MethodInterceptor {
+
+			@Override
+			public Object invoke(MethodInvocation invocation) throws Throwable {
+				callOrder.add("a");
+				latch.countDown();
+				return invocation.proceed();
+			}
+
+		}
+		adviceChain.add(new TestGeneralAdvice());
+
+		class TestSourceAdvice extends AbstractMessageSourceAdvice {
+
+			@Override
+			public boolean beforeReceive(MessageSource<?> target) {
+				callOrder.add("b");
+				latch.countDown();
+				return true;
+			}
+
+			@Override
+			public Message<?> afterReceive(Message<?> result, MessageSource<?> target) {
+				callOrder.add("d");
+				latch.countDown();
+				return result;
+			}
+
+		}
+		adviceChain.add(new TestSourceAdvice());
+
+		adapter.setAdviceChain(adviceChain);
+		adapter.afterPropertiesSet();
+		adapter.start();
+		assertTrue(latch.await(10, TimeUnit.SECONDS));
+		assertThat(callOrder, contains("a", "b", "c", "d")); // advice + advice + source + advice
+		adapter.stop();
+	}
+
+	@Test
+	public void testActiveIdleAdvice() throws Exception {
+		SourcePollingChannelAdapter adapter = new SourcePollingChannelAdapter();
+		final CountDownLatch latch = new CountDownLatch(5);
+		final LinkedList<Long> triggerPeriods = new LinkedList<Long>();
+		final DynamicPeriodicTrigger trigger = new DynamicPeriodicTrigger(10);
+		adapter.setSource(new MessageSource<Object>() {
+
+			@Override
+			public Message<Object> receive() {
+				triggerPeriods.add(trigger.getPeriod());
+				Message<Object> m = null;
+				if (latch.getCount() % 2 == 0) {
+					m = new GenericMessage<Object>("foo");
+				}
+				latch.countDown();
+				return m;
+			}
+		});
+		QueueChannel channel = new QueueChannel();
+		adapter.setOutputChannel(channel);
+		SimpleActiveIdleMessageSourceAdvice toggling = new SimpleActiveIdleMessageSourceAdvice(trigger);
+		toggling.setActivePollPeriod(11);
+		toggling.setIdlePollPeriod(12);
+		adapter.setAdviceChain(Collections.singletonList(toggling));
+		configure(adapter);
+		adapter.afterPropertiesSet();
+		adapter.start();
+		assertTrue(latch.await(10, TimeUnit.SECONDS));
+		adapter.stop();
+		while (triggerPeriods.size() > 5) {
+			triggerPeriods.removeLast();
+		}
+		assertThat(triggerPeriods, Matchers.contains(10L, 12L, 11L, 12L, 11L));
```

Comment:
How can this test show correct results if it isn't supplied with the `DynamicPeriodicTrigger` ? The `new PeriodicTrigger(10)` is by default